### ISSUE TITLE
current_password can't be blank for model update

### ISSUE THREAD

**Issue by augnustin** [danschultzer/pow#430]
So I set up my changeset like this:

```
  def changeset(model, params \\ %{}) do
    model
    |> cast(params, @fields)
    |> pow_changeset(params)
    |> pow_extension_changeset(params)
    |> validate_required([:email])
    |> validate_format(:email, ~r/@/)
    |> unique_constraint(:email)
  end
```

When I try to make an update with something else than password, I get:

```
[error] #Ecto.Changeset<action: :update, changes: %{}, errors: [current_password: {"can't be blank", [validation: :required]}], data: #MyApp.User<>, valid?: false>

# or

[error] #Ecto.Changeset<action: :update, changes: %{address4: "1 example st", city_label: "Example", country_code: "FR", country_label: "France", insee_code: "", postal_code: "XXXXX"}, errors: [current_password: {"can't be blank", [validation: :required]}], data: #MyApp.User<>, valid?: false>
```

What am I doing wrong?

I tried adding `:password`, `password_confirmation` and `current_password` into the `@fields` list, but it doesn't work.

Shouldn't the `pow` changesets be called only if some fields are present (`password`)?

Thanks

**Comment by Schultzer** [comment]
I would disagree on this. `pow_changeset/2` should be as strict as possible.

if your requirements are different, then you should write your own changeset.

**Comment by danschultzer** [comment]
It wouldn't be safe if the client could just remove fields in the request to prevent the fields from being validated. In that case an attacker who have highjacked a session id could just change the password. What I usually recommend is to keep that page to update credentials (email and password), and have a different controller with a profile update page for anything else.

You would set up a custom changeset for that, so it could look like:

```elixir
  def changeset(model, params \\ %{}) do
    model
    |> pow_changeset(params)
    |> pow_extension_changeset(params)
    |> profile_changeset(params)
  end

  def profile_changeset(model, params \\ %{}) do
    cast(model , params, @fields)
  end
```

And then just use `profile_changeset/2` in the profile controller.

Also I removed the email validation as when the user id is email in Pow, it'll handle all validations including formatting: https://hexdocs.pm/pow/1.0.18/Pow.Ecto.Schema.Changeset.html#user_id_field_changeset/3

**Comment by augnustin** [comment]
Not sure I understand.

I'd expect the following behaviour:

```
User.changeset(user, %{password: "new_password", password_confirmation: "new_password", current_password: "exact_current_password"}) # updates password

User.changeset(user, %{address: "address"}) # updates address

User.changeset(user, %{password: "new_password", password_confirmation: "new_password", current_password: "exact_current_password", address: "address"}) # updates password and address

# but

User.changeset(user, %{password: "new_password", password_confirmation: "new_password", current_password: "wrong_current_password"}) # doesn't nothing (wrong current password error)

User.changeset(user, %{password: "new_password", password_confirmation: "new_password"}) # doesn't nothing (blank current password error)

User.changeset(user, %{password: "new_password", password_confirmation: "new_password", current_password: "wrong_current_password", address: "address"}) # does nothing (wrong current password error)

User.changeset(user, %{password: "new_password", password_confirmation: "new_password", address: "address"}) # does nothing (blank current password error)
```

How is that unsafe? I reckon this is how it works with in Rails.

**Comment by augnustin** [comment]
I sorted myself out with the following strategy:

```
  def changeset(model, %{"password" => _pw} = params) do
    model
    |> pow_changeset(params)
    |> changeset(Map.drop(params, ["password", "password_confirmation", "current_password"]))
  end

  def changeset(model, params \\ %{}) do
    model
    |> cast(params, @fields)
    |> pow_extension_changeset(params)
    # and so on ...
  end
```

So that if `password` attribute is set, the pow changeset is called, if not, the "usual" happens.

**Comment by danschultzer** [comment]
> How is that unsafe? I reckon this is how it works with in Rails.

That behavior isn't necessarily unsafe, but the implementation might be. The reason I recommend separating the two concerns is to ensure that you don't accidentally permit changing credentials without authenticating.

> I sorted myself out with the following strategy:

That way it would still require current password since when you submit the form there will always be an empty password attribute. If you check that the password is not an empty string then you should ensure that the user id (in most cases `email`) can't be change without inputting current password. This means that `@fields` should not have `:email`. Ultimately this will be weird UX wise since nothing will happen with the credentials fields unless you input password.

If you insist to keep it all in the same form and permit changes without current password then maybe this would work (untested):

```elixir
  def changeset(model, params) do
    model
    |> maybe_pow_changeset(params)
    |> pow_extension_changeset(params)
    |> cast(params, @fields)
    # ...
  end

  defp maybe_pow_changeset(changeset, params) do
    pow_changeset = pow_changeset(changeset, params)

    case pow_changeset.changes == %{} do
      true  -> changeset
      false -> pow_changeset
    end
  end
```

Basically run the Pow changeset, and then check if there has been any changes to the struct.
